src/crypto/chacha_poly_aead.cpp

```diff
@@ -27,19 +28,52 @@ int timingsafe_bcmp(const unsigned char* b1, const unsigned char* b2, size_t n)
 
 #endif // TIMINGSAFE_BCMP
 
+void ChaCha20ReKey4096::SetKey(const unsigned char* key, size_t keylen) {
+    assert(keylen == 32);
+    m_ctx.SetKey(key, keylen);
+
+    // set initial sequence number
+    m_seqnr = 0;
+    m_ctx.SetIV(m_seqnr);
+
+    // precompute first chunk of keystream
+    m_ctx.Keystream(m_keystream, KEYSTREAM_SIZE);
+    m_keystream_pos = 0;
+}
+
+void ChaCha20ReKey4096::Crypt(const unsigned char* input, unsigned char* output, size_t bytes) {
+    size_t message_pos = 0;
+
+    // TODO: speedup with a block approach (rather then looping over every byte)
+    while (bytes > message_pos) {
+        output[message_pos] = input[message_pos] ^ ReadLE32(&m_keystream[m_keystream_pos]);
+        m_keystream_pos++;
+        message_pos++;
+        if (m_keystream_pos == KEYSTREAM_SIZE-CHACHA20_POLY1305_AEAD_KEY_LEN) {
+            // we reached the end of the keystream
+            // rekey with the remaining and last 32 bytes and precompute the next 4096 bytes
+            m_ctx.SetKey(&m_keystream[m_keystream_pos], CHACHA20_POLY1305_AEAD_KEY_LEN);
+            m_ctx.SetIV(++m_seqnr);
+            m_ctx.Keystream(m_keystream, KEYSTREAM_SIZE);
+            // reset keystream position
+            m_keystream_pos = 0;
+        }
+    }
+}
+
+ChaCha20ReKey4096::~ChaCha20ReKey4096() {
+    memory_cleanse(m_keystream, KEYSTREAM_SIZE);
+}
+
 ChaCha20Poly1305AEAD::ChaCha20Poly1305AEAD(const unsigned char* K_1, size_t K_1_len, const unsigned char* K_2, size_t K_2_len)
 {
     assert(K_1_len == CHACHA20_POLY1305_AEAD_KEY_LEN);
     assert(K_2_len == CHACHA20_POLY1305_AEAD_KEY_LEN);
     m_chacha_main.SetKey(K_1, CHACHA20_POLY1305_AEAD_KEY_LEN);
     m_chacha_header.SetKey(K_2, CHACHA20_POLY1305_AEAD_KEY_LEN);
-
-    // set the cached sequence number to uint64 max which hints for an unset cache.
-    // we can't hit uint64 max since the rekey rule (which resets the sequence number) is 1GB
-    m_cached_aad_seqnr = std::numeric_limits<uint64_t>::max();
 }
 
-bool ChaCha20Poly1305AEAD::Crypt(uint64_t seqnr_payload, uint64_t seqnr_aad, int aad_pos, unsigned char* dest, size_t dest_len /* length of the output buffer for sanity checks */, const unsigned char* src, size_t src_len, bool is_encrypt)
+bool ChaCha20Poly1305AEAD::Crypt(unsigned char* dest, size_t dest_len /* length of the output buffer for sanity checks */, const unsigned char* src, size_t src_len, bool is_encrypt)
```

Comment:
> Or perhaps I am missing a reason the user of the class might need to know the decrypted length.

I was missing the reason: Tried to do this and realized that the interface for `TransportDeserializer` actually requires the client of this class to be aware of the length because the bytes come in as a stream.

Reply:
Yes, I had a similar idea, it's a bit awkward like this. But apparently there's a good reason.

Reply:
Yeah, we need a way to know the boundaries of the messages coming in a single stream.